src/KeywordCalls.jl

```diff
@@ -24,28 +24,53 @@ function kwcallperm(f, keys)
     return σ[π]
 end
 
-# `baseperm[(f, sortedargs::Tuple{Symbol})]` gives the permutation 
-# from sorted arguments to the ordering declared with @kwcall 
+# `baseperm[(f, sortedargs::Tuple{Symbol})]` gives the permutation
+# from sorted arguments to the ordering declared with @kwcall
 const baseperm = Dict()
 
+# can't use gensym() here, as this can lead to collisions because of precompilation
+# so we use a long enough random string (130+ bits of entropy)
+const KEYWORD_CALLS = Symbol("##KeywordCalls-1tqEpmTrQf0b4aJUZscoCc")
+
+function register_calls!(list)
+    for (f, sargs, π) in list
+        baseperm[(f, sargs)] = π
+    end
+end
+
 """
     @kwcall f(b,a,d)
 
 Declares that any call `f(::NamedTuple{N})` with `sort(N) == (:a,:b,:d)`
 should be dispatched to the method already defined on `f(::NamedTuple{(:b,:a,:d)})`
 """
 macro kwcall(call)
-    esc(_kwcall(call))
+    if !isdefined(__module__, KEYWORD_CALLS)
+        @eval __module__ module $KEYWORD_CALLS
+            # Credit to Takafumi Arakaki for the idea of creating a submodule
+            # within modules using KeywordCalls in order to be able to define
+            # an `__init__` function which does the registration
+            # (as we can't directly create/modify `__init__` for the given module).
+            const perm = []
+            __init__() = $register_calls!(perm)
+        end
```

Comment:
Can `KeywordCalls.__init__` create this module? Then we wouldn't need the existence check, and the macro would be a little simpler. Or maybe precompilation still messes things up?

Reply:
> Can KeywordCalls.__init__ create this module?

If you mean create a module within the client package, I don't think so. When KC is precompiled, it doesn't know about client packages. Neither when KC is loaded, because used packages within a client package are loaded before the client.
 
My understanding is that when a client package uses one or more times `@kwcall`, only itself knows about the new elements added to the cache (at precompilation time). How to communicate this stuff to KC? This solution stores this stuff in an array, which persists precompilation. And at load time, via the `__init__` function, it transfers this knowledge to KC via what I called in this PR the `register_calls!` function.  KC has no knowledge about client packages, so I don't see how `KeywordCalls.__init__` could create any module within them. But maybe I misunderstood the question, but I don't see a simpler solution for the moment.